File: JS/Tema9_Ejercicio10.js
/*
 * TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS - EJERCICIO 10
 * =========================================================
 * Sistema completo: Red social con OOP.
 * 
 * EXPLICACIÓN:
 * - Integración de todos los conceptos
 * - Herencia, composición, encapsulación
 * - Sistema funcional completo
 */

class Usuario {
    #contraseña = "";
    #amigos = [];
    
    constructor(nombre, email, contraseña) {
        this.nombre = nombre;
        this.email = email;
        this.#contraseña = contraseña;
        this.publicaciones = [];
    }
    
    publicar(contenido) {
        const post = {
            id: this.publicaciones.length + 1,
            contenido,
            fecha: new Date().toLocaleDateString(),
            likes: 0
        };
        this.publicaciones.push(post);
    }
    
    agregarAmigo(usuario) {
        if (!this.#amigos.includes(usuario)) {
            this.#amigos.push(usuario);
        }
    }
    
    get amigos() {
        return this.#amigos.map(a => a.nombre);
    }
    
    set contraseña(nueva) {
        if (nueva.length >= 6) {
            this.#contraseña = nueva;
        }
    }
    
    obtenerPerfil() {
        return `${this.nombre} (${this.email}) - ${this.#amigos.length} amigos`;
    }
}

class Publicacion {
    constructor(usuario, contenido) {
        this.usuario = usuario;
        this.contenido = contenido;
        this.fecha = new Date().toLocaleDateString();
        this.likes = 0;
    }
    
    darLike() {
        this.likes++;
    }
    
    obtenerInfo() {
        return `${this.usuario.nombre}: "${this.contenido}" (${this.likes} likes)`;
    }
}

class RedSocial {
    constructor(nombre) {
        this.nombre = nombre;
        this.usuarios = new Map();
    }
    
    registrarUsuario(nombre, email, contraseña) {
        const usuario = new Usuario(nombre, email, contraseña);
        this.usuarios.set(nombre, usuario);
        return usuario;
    }
    
    conectarAmigos(nombre1, nombre2) {
        const u1 = this.usuarios.get(nombre1);
        const u2 = this.usuarios.get(nombre2);
        if (u1 && u2) {
            u1.agregarAmigo(u2);
            u2.agregarAmigo(u1);
        }
    }
    
    mostrarInfoUsuario(nombre) {
        const usuario = this.usuarios.get(nombre);
        console.log(`\n${usuario.obtenerPerfil()}`);
        console.log(`Amigos: ${usuario.amigos.join(", ") || "Sin amigos"}`);
        console.log(`Publicaciones: ${usuario.publicaciones.length}`);
        usuario.publicaciones.forEach(p => {
            console.log(`  - ${p.contenido}`);
        });
    }
}

console.log("=== RED SOCIAL ===\n");

const red = new RedSocial("TechNet");

const juan = red.registrarUsuario("Juan", "juan@email.com", "password123");
const ana = red.registrarUsuario("Ana", "ana@email.com", "secure456");
const carlos = red.registrarUsuario("Carlos", "carlos@email.com", "pass789");

juan.publicar("¡Hola a todos!");
juan.publicar("Aprendiendo JavaScript OOP");

ana.publicar("Primer día en TechNet");

red.conectarAmigos("Juan", "Ana");
red.conectarAmigos("Juan", "Carlos");
red.conectarAmigos("Ana", "Carlos");

red.mostrarInfoUsuario("Juan");
red.mostrarInfoUsuario("Ana");
red.mostrarInfoUsuario("Carlos");

/*
 * SALIDA ESPERADA:
 * === RED SOCIAL ===
 * 
 * Juan (juan@email.com) - 2 amigos
 * Amigos: Ana, Carlos
 * Publicaciones: 2
 *   - ¡Hola a todos!
 *   - Aprendiendo JavaScript OOP
 * 
 * Ana (ana@email.com) - 2 amigos
 * Amigos: Juan, Carlos
 * Publicaciones: 1
 *   - Primer día en TechNet
 * 
 * Carlos (carlos@email.com) - 2 amigos
 * Amigos: Juan, Ana
 * Publicaciones: 0
 */

File: Python/Tema9_Ejercicio10.py
"""
TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS
EJERCICIO 10: RED SOCIAL (INTEGRACIÓN)
======================================

Sistema completo integrando todos los conceptos
"""

# RED SOCIAL
print("=== RED SOCIAL ===\n")

from datetime import datetime

class Usuario:
    def __init__(self, nombre, email):
        self.nombre = nombre
        self.email = email
        self._amigos = set()
        self.publicaciones = []
    
    def agregar_amigo(self, usuario):
        self._amigos.add(usuario.nombre)
        print(f"  {self.nombre} y {usuario.nombre} son amigos")
    
    def obtener_amigos(self):
        return list(self._amigos)
    
    def publicar(self, contenido):
        pub = {"contenido": contenido, "fecha": datetime.now()}
        self.publicaciones.append(pub)
        print(f"  {self.nombre} publicó: {contenido}")

class Publicacion:
    def __init__(self, autor, contenido):
        self.autor = autor
        self.contenido = contenido
        self.likes = 0
        self.comentarios = []
    
    def dar_like(self):
        self.likes += 1
    
    def comentar(self, comentario):
        self.comentarios.append(comentario)

class RedSocial:
    def __init__(self, nombre):
        self.nombre = nombre
        self.usuarios = {}
    
    def registrar_usuario(self, nombre, email):
        if nombre not in self.usuarios:
            self.usuarios[nombre] = Usuario(nombre, email)
            print(f"  Usuario '{nombre}' registrado")
            return self.usuarios[nombre]
        return None
    
    def conectar_usuarios(self, nombre1, nombre2):
        if nombre1 in self.usuarios and nombre2 in self.usuarios:
            self.usuarios[nombre1].agregar_amigo(self.usuarios[nombre2])
            self.usuarios[nombre2].agregar_amigo(self.usuarios[nombre1])
    
    def obtener_usuario(self, nombre):
        return self.usuarios.get(nombre)

# CREAR RED SOCIAL
print(f"Creando '{usuario}\n")

red = RedSocial("Mi Red Social")

# Registrar usuarios
juan = red.registrar_usuario("Juan", "juan@email.com")
ana = red.registrar_usuario("Ana", "ana@email.com")
carlos = red.registrar_usuario("Carlos", "carlos@email.com")

# Conectar usuarios
print("\nConectando usuarios:")
red.conectar_usuarios("Juan", "Ana")
red.conectar_usuarios("Juan", "Carlos")
red.conectar_usuarios("Ana", "Carlos")

# Publicar
print("\nPublicando contenido:")
juan.publicar("¡Hola a todos!")
ana.publicar("¿Cómo están?")
carlos.publicar("Saludos desde Madrid")

# Mostrar información
print("\nInformación de usuarios:")
for nombre in ["Juan", "Ana", "Carlos"]:
    usuario = red.obtener_usuario(nombre)
    print(f"  {usuario.nombre}:")
    print(f"    Email: {usuario.email}")
    print(f"    Amigos: {usuario.obtener_amigos()}")
    print(f"    Publicaciones: {len(usuario.publicaciones)}")

"""
SALIDA ESPERADA:
=== RED SOCIAL ===

Creando red social

  Usuario 'Juan' registrado
  Usuario 'Ana' registrado
  Usuario 'Carlos' registrado

Conectando usuarios:
  Juan y Ana son amigos
  Ana y Juan son amigos
  Juan y Carlos son amigos
  Carlos y Juan son amigos
  Ana y Carlos son amigos
  Carlos y Ana son amigos

Publicando contenido:
  Juan publicó: ¡Hola a todos!
  Ana publicó: ¿Cómo están?
  Carlos publicó: Saludos desde Madrid

Información de usuarios:
  Juan:
    Email: juan@email.com
    Amigos: ['Ana', 'Carlos']
    Publicaciones: 1
...
"""
